src/foam/lib/json/MultiLineStringParser.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.lib.json;
+
+import foam.lib.parse.PStream;
+import foam.lib.parse.Parser;
+import foam.lib.parse.ParserContext;
+
+public class MultiLineStringParser
+  implements Parser
+{
+  protected final static char ESCAPE = '\\';
+  protected final static ThreadLocal<Parser> unicodeParser = ThreadLocal.withInitial(UnicodeParser::new);
+  protected final static ThreadLocal<Parser> asciiEscapeParser = ThreadLocal.withInitial(ASCIIEscapeParser::new);
+  protected final static ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
+
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder b = super.get();
+      b.setLength(0);
+      return b;
+    }
+  };
+
+  public MultiLineStringParser() {
+  }
+
+  public PStream parse(PStream ps, ParserContext x) {
+    if ( ! ps.valid() ) return null;
+    char delim = ps.head();
+    char lastc = delim;
+
+    for( int i = 0; i < 3; i++ ) {
+      if ( delim != '"' ) return null;
+      ps = ps.tail();
+      delim = ps.head();
+    }
+
+    delim = lastc;
+
+    StringBuilder builder = sb.get();
+
+    while ( ps.valid() ) {
+      char c = ps.head();
+
+      if ( c == delim && lastc != ESCAPE && ps.tail().head() == delim && ps.tail().tail().head() == delim ) break;
+
+      PStream tail = ps.tail();
+
+      if ( c == ESCAPE ) {
+        char   nextChar        = ps.tail().head();
```

Comment:
Seems like a lot of duplicated logic between the Multiline and the Singeline parser.

This seems a lot like just:

`Sequence(Literal("\"\"\""), Repeat(Not(Literal("\"\"\""), Alt(AsciiEscape, UnicodeEscape, AnyChar()))), Literal("\"\"\""))`

Rather than implementing the entire loop itself.

Upon further consideration though, I don't think our current StringParser even rejects newlines in strings.  I don't think there's any need for Multiline vs Sinlgeline parsers, the only real different is whether the string is delimited with a single quotation mark or a triple quotation mark so that people don't have to escape every quotation mark in the string.

If we tweak the existing StringParser to check if the string starts with a triple " in addition to the current checks for a single " or ', then we I think we can just continue to use that rather than adding new parser.

Reply:
In StringParser.java, instead of using a single char as a delim we can use a parser similar to how Repeat is implemented.  Something like

```
public class StringParser {
  private Parser delimiterParser = new Alt(new Literal("\""),
                                           new Literal("'"),
                                           new Literal("\"\"\""));
                                     
  public PStream parse(PStream ps, ParserContext x) {
    ps = ps.apply(delimiterParser, x);
    if ( ! ps ) return null;

    Parser delimiter = new Literal((String)ps.getValue());

    PStream result;
    while ( ps.valid() ) {
      result = ps.apply(delimiter, x);
      if ( result != null ) break;
      ...
    }
  }
}
```